index/scorch/scorch.go

```diff
@@ -515,6 +515,26 @@ func (s *Scorch) diskFileStats(rootSegmentPaths map[string]struct{}) (uint64,
 	return numFilesOnDisk, numBytesUsedDisk, numBytesOnDiskByRoot
 }
 
+// reClaimableDocsRatio gives a ratio about the obsoleted or
+// reclaimable documents present in a scorch index.
+func (s *Scorch) reClaimableDocsRatio() float64 {
+	var totalCount, liveCount uint64
+	s.rootLock.RLock()
+	indexSnapshot := s.root
+	s.rootLock.RUnlock()
+	for _, segmentSnapshot := range indexSnapshot.segment {
+		if _, ok := segmentSnapshot.segment.(segment.PersistedSegment); ok {
+			totalCount += uint64(segmentSnapshot.FullSize())
+			liveCount += uint64(segmentSnapshot.Count())
+		}
+	}
+
+	if totalCount > 0 {
+		return float64(totalCount-liveCount) / float64(totalCount)
+	}
+	return 0
+}
+
 func (s *Scorch) rootDiskSegmentsPaths() map[string]struct{} {
```

Comment:
As part of this change, would you also move the acquiring of RLock into this function?
I feel it's better readable here than in the calling function like we do in StatsMap() API below - avoids any future mistakes.

Reply:
hm.. actually I feel the otherway.. the lock should be at the StatsMap() as that assures that the whole op is atomic..
Else it is not guaranteed from a single index snapshot.

So, let me take out the lock and call from parent like this rootDiskSegmentsPaths.
It has its own ugliness... :)

Reply:
Seems like if we have multiple stats using the current snapshot, we should get the lock once, grab a copy of the pointer to that snapshot, and pass it around, so we minimize lock contention.

Reply:
hm.. not sure whether I am overlooking something here.. But will a copy of index snapshot works? 
Since if the root changes, then there is a chance that disk files/stats becomes obsolete due to clean ups.. 
So looks like we need to readLock until the disk stats are taken?
(i am on a friday beer)

Reply:
> hm.. actually I feel the otherway.. the lock should be at the StatsMap() as that assures that the whole op is atomic..
> Else it is not guaranteed from a single index snapshot.
> 
> So, let me take out the lock and call from parent like this rootDiskSegmentsPaths.
> It has its own ugliness... :)

Ok in that case, per our naming - suffix LOCKED to these function names then?

Reply:
I don't think we should hold the lock, I think we just have to also increment the refcount.  In Bluge, I cleaned up scorch to always go through this helper, because I didn't want anyone touching `s.root` directly:

```
func (s *Writer) currentSnapshot() *Snapshot {
	s.rootLock.RLock()
	var rv *Snapshot
	if s.root != nil {
		rv = s.root
		if rv != nil {
			rv.addRef()
		}
	}
	s.rootLock.RUnlock()
	return rv
}
```

Obviously some of the type names have changed, but you can see the pattern.  Anyone that wants to access the current snapshot, goes through this method, and simply calls Close() when done.

Reply:
good idea!, utility is already there. Just need to use it!